daml-script/export/src/main/scala/com/daml/script/export/TreeUtils.scala

```diff
@@ -163,17 +163,24 @@ object TreeUtils {
         }
   }
 
-  case class CreatedContract(cid: ContractId, tplId: TemplateId, path: List[Selector])
+  trait CreatedContract {
+    def cid: ContractId
+    def tplId: TemplateId
+  }
+
+  case class SimpleCreatedContract(cid: ContractId, tplId: TemplateId) extends CreatedContract
+  case class CreatedContractWithPath(cid: ContractId, tplId: TemplateId, path: List[Selector])
+      extends CreatedContract
```

Comment:
In general, you want all case classes to be final. This is done to stop case classes being extended (which the language allows). The problem with extending case classes is that careless extensions can result in equality and friends being broken (with often surprising results!), and so this is best avoided by mortals.

That said, maybe you want `CreatedContract` to be a sealed trait? Suggesting this as it would allow you to pattern match on the types of created contract in case statements, and that might be a convenience for your expected use cases?

Reply:
thanks @carlpulley-da, I understand better now :)

I've made the `case class`es final and the `trait` `sealed` now. I don't really need to pattern match on `SimpleCreatedContract` vs `CreatedContractWithPath` since my functions either care only about `CreatedContractWithPath`s or only about the common fields.

Additionally, since I didn't like the name `SimpleCreatedContract`, I moved it inside a new `object CreatedContract` and used it to implement an `apply` method, so now I just need `CreatedContract(...)` at the use sites.